debian/cloud-init.postinst

```diff
@@ -276,6 +276,51 @@ fix_azure_upgrade_1611074() {
     rm "$tmpf" || :
 }
 
+fix_lp1889555() {
+    local oldver="$1" last_bad_ver="20.2-94-g3d06abc2-0ubuntu1"
+    dpkg --compare-versions "$oldver" le-nl "$last_bad_ver" || return 0
+
+    # if cloud-init's grub module did not run, then it did not break anything.
+    [ -f /var/lib/cloud/instance/sem/config_grub_dpkg ] || return 0
+
+    # Don't do anything unless we have grub
+    [ -x /usr/sbin/grub-install ] || return 0
+
+    # Make sure that we are not chrooted.
+    [ "$(stat -c %d:%i /)" != "$(stat -c %d:%i /proc/1/root/.)" ] && return 0
+
+    # Check if we are in a container, i.e. LXC
+    if systemd-detect-virt --quiet --container || lxc-is-container 2>/dev/null; then
+        return 0
+    fi
+
+    # This bug only applies to NVMe devices
+    [ -e /dev/nvme0 ] || return 0
```

Comment:
This bug isn't limited to NVMe devices only, its simply the most common failure mode (on cloud providers). If your system didn't have any of the original hard coded directories, e.g. /dev/sda, /dev/vda, /dev/xvda, /dev/sda1, /dev/vda1 or /dev/xvda1 then it defaulted to /dev/sda. 

If my system used IDE controllers, with /dev/hda being my root disk, I would be in the same boat as /dev/nvme* users, since /dev/sda would be selected. It's why we moved to the generic grub-probe solution, over simply hardcoding /dev/nvme0. 

Note also, that /dev/nvme0 might not be the boot disk, it could be the scratch disk on a system with multiple NVMe devices, as the kernel NVMe device ordering is not stable, hence the push for /dev/disk/by-id in the refactor.

I know its risky, but consider removing this check.

Reply:
> If my system used IDE controllers, with /dev/hda being my root disk, I would be in the same boat as /dev/nvme* users, since /dev/sda would be selected. It's why we moved to the generic grub-probe solution, over simply hardcoding /dev/nvme0.

Right, the intent here is to fix a class of cloud instances where we can be pretty confident that we will be doing the right thing. AFAIK, we do not have a similar class of instances which use `/dev/hda` (or anything other than the previously-hard-coded values or NVMe, for that matter).

> Note also, that /dev/nvme0 might not be the boot disk, it could be the scratch disk on a system with multiple NVMe devices, as the kernel NVMe device ordering is not stable, hence the push for /dev/disk/by-id in the refactor.

`/dev/nvme0` certainly won't be the boot disk, that would be `/dev/nvme0n1` even on single-disk instances; this is pedantic, but my point is that this check is intended to ensure we only run on systems which have any NVMe disks; we will still use the logic in `cc_grub_dpkg` to determine the boot disk (whether or not it is NVMe) and use the `by-id` symlink.